platform/linux-generic/odp_packet_io.c

```diff
@@ -2472,3 +2485,345 @@ int odp_pktout_ts_read(odp_pktio_t hdl, odp_time_t *ts)
+{
+	lso_profile_t *lso_prof = (lso_profile_t *)(uintptr_t)lso_profile;
+
+	if (lso_profile == ODP_LSO_PROFILE_INVALID || lso_prof->used == 0) {
+		ODP_ERR("Bad handle\n");
+		return -1;
+	}
+
+	odp_spinlock_lock(&pktio_global->lock);
+	lso_prof->used = 0;
+	pktio_global->num_lso_profiles--;
+	odp_spinlock_unlock(&pktio_global->lock);
+
+	return 0;
+}
+
+int odp_packet_lso_request(odp_packet_t pkt, const odp_packet_lso_opt_t *lso_opt)
+{
+	odp_packet_hdr_t *pkt_hdr = packet_hdr(pkt);
+	lso_profile_t *lso_prof = (lso_profile_t *)(uintptr_t)lso_opt->lso_profile;
+
+	if (lso_opt->lso_profile == ODP_LSO_PROFILE_INVALID || lso_prof->used == 0) {
+		ODP_ERR("Bad LSO profile handle\n");
+		return -1;
+	}
+
+	pkt_hdr->p.flags.lso     = 1;
+	pkt_hdr->lso_max_payload = lso_opt->max_payload_len;
+	pkt_hdr->lso_profile_idx = lso_prof->index;
+	odp_packet_payload_offset_set(pkt, lso_opt->payload_offset);
+
+	return 0;
+}
+
+static int lso_update_custom(lso_profile_t *lso_prof, odp_packet_t pkt, int segnum)
+{
+	void *ptr;
+	int i, mod_op;
+	uint32_t offset;
+	uint8_t size;
+	int num_custom = lso_prof->param.custom.num_custom;
+	uint64_t u64 = 0;
+	uint32_t u32 = 0;
+	uint16_t u16 = 0;
+	uint8_t  u8 = 0;
+
+	for (i = 0; i < num_custom; i++) {
+		mod_op = lso_prof->param.custom.field[i].mod_op;
+		offset = lso_prof->param.custom.field[i].offset;
+		size   = lso_prof->param.custom.field[i].size;
+
+		if (size == 8)
+			ptr = &u64;
+		else if (size == 4)
+			ptr = &u32;
+		else if (size == 2)
+			ptr = &u16;
+		else
+			ptr = &u8;
+
+		if (odp_packet_copy_to_mem(pkt, offset, size, ptr)) {
+			ODP_ERR("Read from packet failed at offset %u\n", offset);
+			return -1;
+		}
+
+		if (size == 8)
+			u64 = odp_be_to_cpu_64(u64);
+		else if (size == 4)
+			u32 = odp_be_to_cpu_32(u32);
+		else if (size == 2)
+			u16 = odp_be_to_cpu_16(u16);
+
+		if (mod_op == ODP_LSO_ADD_SEGMENT_NUM) {
+			if (size == 8)
+				u64 += segnum;
+			else if (size == 4)
+				u32 += segnum;
+			else if (size == 2)
+				u16 += segnum;
+			else
+				u8  += segnum;
+		}
+
+		if (size == 8)
+			u64 = odp_cpu_to_be_64(u64);
+		else if (size == 4)
+			u32 = odp_cpu_to_be_32(u32);
+		else if (size == 2)
+			u16 = odp_cpu_to_be_16(u16);
+
+		if (odp_packet_copy_from_mem(pkt, offset, size, ptr)) {
+			ODP_ERR("Write to packet failed at offset %u\n", offset);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int pktout_send_lso(odp_pktout_queue_t queue, odp_packet_t packet,
+			   const odp_packet_lso_opt_t *lso_opt)
+{
+	odp_pool_t pool;
+	odp_packet_t pkt;
+	int i, ret, num, num_pkt, num_full;
+	uint32_t hdr_len, pkt_len, pkt_payload, payload_len, left_over_len, offset;
+	int left_over = 0;
+	int num_free = 0;
+	int first_free = 0;
+	odp_lso_profile_t lso_profile = lso_opt->lso_profile;
+	lso_profile_t *lso_prof = (lso_profile_t *)(uintptr_t)lso_profile;
+	int num_custom = lso_prof->param.custom.num_custom;
+
+	payload_len = lso_opt->max_payload_len;
+	hdr_len     = lso_opt->payload_offset;
```

Comment:
It could be worth it to add assert checks for lso_opt.max_payload_len and lso_opt.payload_offset if this doesn't create extra overhead.

Reply:
fixed